### PR TITLE
ESQL: Begin optimizing `Block#lookup`

### PR DESCRIPTION
**Pull request by nik9000**
This creates the infrastructure to allow optimizing the `lookup` method when applied to `Vector`s and then implements that optimization for constant vectors. Constant vectors now take one of six paths:
1. An empty positions `Block` yields an empty result set.
2. If `positions` is a `Block`, perform the un-optimized lookup.
3. If the `min` of the `positions` *Vector* is less that 0 then throw an exception.
4. If the `min` of the positions Vector is greater than the number of positions in the lookup block then return a single `ConstantNullBlock` because you are looking up outside the range.
5. If the `max` of the positions Vector is less than the number of positions in the lookup block then return a `Constant$Type$Block` with the same value as the lookup block. This is a lookup that's entirely within range.
6. Otherwise return the unoptimized lookup.

This is *fairly* simple but demonstrates how we can plug in more complex optimizations later.

### CODE REVIEW

File: libs/core/src/main/java/org/elasticsearch/core/ReleasableIterator.java [hunk 1]
```diff
@@ -46,4 +46,29 @@ public String toString() {
...
+    static <T extends Releasable> ReleasableIterator<T> empty() {
+        return new ReleasableIterator<>() {
+            @Override
+            public boolean hasNext() {
+                return false;
+            }
+
+            @Override
+            public T next() {
+                return null;
```
Comment (rjernst) [inline libs/core/src/main/java/org/elasticsearch/core/ReleasableIterator.java:62]: We could assert false here? next should never be called in this case

**Review by rjernst — APPROVED** [review]
Core lib change looks good to me, one small suggestion.

File: libs/core/src/main/java/org/elasticsearch/core/ReleasableIterator.java (hunk 1 above)
Comment (nik9000) [reply libs/core/src/main/java/org/elasticsearch/core/ReleasableIterator.java:62]: Sounds good!

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Nik!

**Pull request merged**